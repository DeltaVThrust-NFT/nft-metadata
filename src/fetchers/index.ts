import { fetchUnderlyingContractData } from './defaultFetcher'
import { getAddress } from '@ethersproject/address'

export function fetcherLookup(contractAddress: string) {
  const safeAddress = getAddress(contractAddress)
  switch (safeAddress) {
    default:
      return fetchUnderlyingContractData
  }
}
