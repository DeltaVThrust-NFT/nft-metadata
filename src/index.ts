export * from './constants'
export * from './metadata'
export * from './agent'
export * from './uri'
