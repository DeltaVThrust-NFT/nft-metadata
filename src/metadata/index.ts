export * from './fetcher'
export * from './hashmasks'
export * from './loot'
export * from './normaliser'
export * from './zora'
