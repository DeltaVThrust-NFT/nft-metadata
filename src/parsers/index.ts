import { getAddress } from '@ethersproject/address'
import { parseGenericMetadata } from './openseaMetadataParser'
import {
  ART_BLOCKS_CURATED_TOKEN_ADDRESS,
  ART_BLOCKS_TOKEN_ADDRESS,
  HASHMASKS_TOKEN_ADDRESS,
  MAKERSPLACE_TOKEN_ADDRESS,
  ZORA_TOKEN_ADDRESS,
} from '../constants'
import { parseZoraMetadata } from './zoraMetadataParser'
import { parseMakersplaceMetadata } from './makersplaceMetadataParser'
import { parseHashmasksMetadata } from './hashmasksMetadataParser'
import { parseArtblocksMetadata } from './artblocksMetadataParser'

export function parserLookup(contractAddress: string) {
  const safeAddress = getAddress(contractAddress)
  switch (safeAddress) {
    case ART_BLOCKS_TOKEN_ADDRESS:
      return parseArtblocksMetadata
    case ART_BLOCKS_CURATED_TOKEN_ADDRESS:
      return parseArtblocksMetadata
    case HASHMASKS_TOKEN_ADDRESS:
      return parseHashmasksMetadata
    case MAKERSPLACE_TOKEN_ADDRESS:
      return parseMakersplaceMetadata
    case ZORA_TOKEN_ADDRESS:
      return parseZoraMetadata
    default:
      return parseGenericMetadata
  }
}
