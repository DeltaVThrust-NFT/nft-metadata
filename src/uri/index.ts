export { getAlternateContractCall } from './contract'
export {
  fetchMimeType,
  fetchURI,
  fetchWithRetries,
  fetchWithTimeout,
  parseDataUri,
} from './fetch'
export { isIPFS, getIPFSUrl, getPrivateGateway } from './ipfs'
export { getStaticURI, getURIData } from './static'
