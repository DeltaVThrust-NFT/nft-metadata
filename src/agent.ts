import { JsonRpcProvider } from '@ethersproject/providers'
import { Networkish } from '@ethersproject/networks'
import {
  fetchMimeType,
  fetchURI,
  getAlternateContractCall,
  getIPFSUrl,
  getPrivateGateway,
  getStaticURI,
  getURIData,
  IPFS_IO_GATEWAY,
} from './uri'
import { getAddress } from '@ethersproject/address'
import { Erc721Factory } from '@zoralabs/core/dist/typechain'
import { fetchOnChainData, normaliseURIData } from './metadata'

type AgentOptions = {
  network: Networkish
  networkUrl: string
  ipfsGatewayUrl?: string
  timeout?: number
}

export interface NftMetadata {
  tokenId: string
  tokenAddress: string
  metadata: any
  tokenURI: string
  tokenURL: string
  tokenURLMimeType: string
  name?: string
  description?: string
  contentURL?: string
  contentURLMimeType?: string
  imageURL?: string
  imageURLMimeType?: string
  externalLink?: string
  attributes?: Record<string, any>[]
}

export class Agent {
  timeout: number
  ipfsGatewayUrl: string
  provider: JsonRpcProvider

  constructor(options: AgentOptions) {
    this.provider = new JsonRpcProvider(options.networkUrl, options.network)
    this.ipfsGatewayUrl = options.ipfsGatewayUrl || IPFS_IO_GATEWAY
    this.timeout = options.timeout || 10000
  }

  public async fetchTokenURI(tokenAddress: string, tokenId: string) {
    const staticURI = getStaticURI(tokenAddress, tokenId)
    if (staticURI) {
      return staticURI
    }
    const alternateMethod = getAlternateContractCall(
      tokenAddress,
      tokenId,
      this.provider,
    )
    if (alternateMethod) {
      return alternateMethod
    }
    const contract = Erc721Factory.connect(tokenAddress, this.provider)
    try {
      return contract.tokenURI(tokenId)
    } catch (e) {
      return
    }
  }

  public async fetchURIData(
    tokenAddress: string,
    tokenId: string,
    tokenURI: string,
    ipfsGateway: string,
  ) {
    const alternateMethod = getURIData(tokenAddress, tokenId)
    if (alternateMethod) {
      return alternateMethod
    }
    const resp = await fetchURI(
      tokenURI,
      { timeout: this.timeout },
      ipfsGateway,
    )
    if (!resp) {
      throw new Error(`Failed to fetch uri data for token from: ${tokenURI}`)
    }
    return resp
  }

  private async parseURIData(
    tokenAddress: string,
    tokenId: string,
    tokenURI: string,
    uriData: any,
    ipfsGateway: string,
  ) {
    const onChainData = await fetchOnChainData(
      tokenAddress,
      tokenId,
      this.provider,
    )
    const meta = normaliseURIData(tokenAddress, {
      ...uriData,
      ...onChainData,
      ...(uriData.mimeType && {
        contentURLMimeType: uriData.mimeType,
      }),
    })

    if (meta.image) {
      meta.imageURL = getIPFSUrl(meta.image, ipfsGateway)
    }

    if (meta.animation_url) {
      meta.contentURL = getIPFSUrl(meta.animation_url, ipfsGateway)
    }

    if (!meta.contentURL && meta.imageURL) {
      meta.contentURL = meta.imageURL
    }

    if (meta.contentURL && !meta.contentURLMimeType) {
      meta.contentURLMimeType = await fetchMimeType(meta.contentURL, {
        timeout: this.timeout,
      })
    }

    if (meta.imageURL && !meta.imageURLMimeType) {
      meta.imageURLMimeType = await fetchMimeType(meta.imageURL, {
        timeout: this.timeout,
      })
    }

    const {
      name,
      description,
      attributes,
      external_url: externalURL,
      imageURL,
      imageURLMimeType,
      contentURL,
      contentURLMimeType,
    } = meta

    return {
      tokenURL: getIPFSUrl(tokenURI, ipfsGateway),
      tokenURLMimeType: 'application/json',
      ...(name && { name }),
      ...(description && { description }),
      ...(imageURL && { imageURL }),
      ...(imageURLMimeType && { imageURLMimeType }),
      ...(contentURL && { contentURL }),
      ...(contentURLMimeType && { contentURLMimeType }),
      ...(attributes && { attributes }),
      ...(externalURL && { externalURL }),
    }
  }

  public async fetchMetadata(
    rawAddress: string,
    tokenId: string,
  ): Promise<NftMetadata> {
    const tokenAddress = getAddress(rawAddress)
    const tokenURI = await this.fetchTokenURI(tokenAddress, tokenId)
    if (!tokenURI) {
      throw new Error(
        `Failed to get tokenURI token: ${tokenAddress} is unsupported by @zoralabs/nft-metadata`,
      )
    }
    console.log('fetched uri ', { tokenURI })
    const ipfsGateway = getPrivateGateway(tokenAddress) || this.ipfsGatewayUrl
    const URIData = await this.fetchURIData(
      tokenAddress,
      tokenId,
      tokenURI,
      ipfsGateway,
    )
    console.log('fetched uri data: ', { URIData })
    const metadata = await this.parseURIData(
      tokenAddress,
      tokenId,
      tokenURI,
      URIData,
      ipfsGateway,
    )
    console.log('parsed metada: ', { metadata })

    return {
      tokenId,
      tokenAddress,
      metadata: URIData,
      tokenURI,
      ...metadata,
    }
  }
}
