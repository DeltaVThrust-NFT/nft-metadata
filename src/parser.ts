import { JsonRpcProvider } from '@ethersproject/providers'
import { getAddress } from '@ethersproject/address'
import { fetcherLookup } from './fetchers'
import { parserLookup } from './parsers'

export interface NftMetadata {
  metadata: any

  name?: string
  description?: string

  ownerAddress: string

  tokenURI: string
  contentURI: string

  externalLink?: string
  attributes?: Record<string, any>[]
}

export class Parser {
  constructor(private readonly provider: JsonRpcProvider) {}

  public async fetchUnderlyingContractData(
    contractAddress: string,
    tokenId: string,
  ) {
    const fetcher = fetcherLookup(contractAddress)
    return fetcher(this.provider, contractAddress, tokenId)
  }

  public async parseTokenMetadata(
    contractAddress: string,
    tokenId: string,
    tokenURI: string,
  ) {
    const parser = parserLookup(contractAddress)
    return parser(this.provider, contractAddress, tokenId, tokenURI)
  }

  public async fetchAndParseTokenMeta(
    tokenAddress: string,
    tokenId: string,
  ): Promise<NftMetadata> {
    const safeAddress = getAddress(tokenAddress)

    const { tokenURI, ownerAddress } = await this.fetchUnderlyingContractData(
      safeAddress,
      tokenId,
    )

    const parsedMetadata = await this.parseTokenMetadata(
      safeAddress,
      tokenId,
      tokenURI,
    )

    return {
      tokenURI,
      ownerAddress,
      ...parsedMetadata,
    }
  }
}
