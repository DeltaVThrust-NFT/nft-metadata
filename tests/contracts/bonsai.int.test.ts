import { BONSAI_TOKEN_ADDRESS, Agent, Network } from '../../src'
import METADATA_STUB from '../mock-reponses/contracts/bonsai/100.json'
import { testProvider } from '../setupProvider'
import { isAddress } from '@ethersproject/address'

const BONSAI_CRITERIA = {
  input: {
    tokenId: '100',
    tokenAddress: BONSAI_TOKEN_ADDRESS,
  },
  output: {
    metadata: METADATA_STUB,
    name: METADATA_STUB.name,
    description: METADATA_STUB.description,
    tokenURL:
      'https://dweb.link/ipfs/QmXPSFmFqfDTMbLePfGTuYa2Vm9CoqsU11ypiMm1nKL8V9/100',
    tokenURLMimeType: 'application/json',
    contentURL: METADATA_STUB.animation_url,
    contentURLMimeType: 'video/mp4',
    imageURL: METADATA_STUB.image,
    imageURLMimeType: 'image/png',
    attributes: METADATA_STUB.attributes,
  },
}

describe('Bonsai ERC721', () => {
  const parser = new Agent({
    providers: {
      [Network.MAINNET]: testProvider,
    },
    ipfsGateway: 'https://dweb.link',
    fetchTimeout: 15000,
  })

  beforeEach(() => {
    jest.setTimeout(60000)
  })

  it(`should be able to fetch and parse metadata for token id: ${BONSAI_CRITERIA.input.tokenId}`, async () => {
    const { ownerAddress, ...meta } = await parser.fetchAndParseTokenData(
      1,
      BONSAI_CRITERIA.input.tokenAddress,
      BONSAI_CRITERIA.input.tokenId,
    )
    expect(meta).toStrictEqual(BONSAI_CRITERIA.output)
    expect(isAddress(ownerAddress)).toBeTruthy()
  })
})
