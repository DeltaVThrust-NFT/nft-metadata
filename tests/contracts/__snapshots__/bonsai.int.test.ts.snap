// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Bonsai ERC721 should be able to fetch and parse metadata for token id: 100 on network 1 1`] = `
Object {
  "attributes": Array [
    Object {
      "trait_type": "Pot",
      "value": "Yellow Plastic",
    },
    Object {
      "trait_type": "Pet",
      "value": "None",
    },
    Object {
      "trait_type": "Bark",
      "value": "Birch",
    },
    Object {
      "trait_type": "Leaves",
      "value": "Momo Maple",
    },
    Object {
      "trait_type": "Fruit",
      "value": "None",
    },
    Object {
      "trait_type": "Number of fruit",
      "value": "0",
    },
    Object {
      "trait_type": "Ground cover",
      "value": "None",
    },
    Object {
      "trait_type": "Background",
      "value": "day",
    },
  ],
  "contentURL": "https://ipfs.io/ipfs/QmWS953fEQgQo77n7MCkmUYWP76X4mTZuq7pH6MUP2tJs8",
  "contentURLMimeType": "video/mp4",
  "description": "[Image (.png)](https://ipfs.io/ipfs/QmSckSnuJUwn74ZFvmNMob9dTwfZzjnXVyeKxuDYah93hx)

[Animation (.mp4)](https://ipfs.io/ipfs/QmWS953fEQgQo77n7MCkmUYWP76X4mTZuq7pH6MUP2tJs8)

[3D Model (.glb)](https://ipfs.io/ipfs/QmS8VGJL1P9kcGb2JNkPyUesYamCLRDPBg8aVuCr9qCDZM)

[AR Object (.usdz)](https://ipfs.io/ipfs/Qmd6rodUdPr3EZ2SfZ2DxjukQUjz5WaSHLBkrjAnW6yH83)
",
  "imageURL": "https://ipfs.io/ipfs/QmSckSnuJUwn74ZFvmNMob9dTwfZzjnXVyeKxuDYah93hx",
  "imageURLMimeType": "image/png",
  "metadata": Object {
    "animation_url": "https://ipfs.io/ipfs/QmWS953fEQgQo77n7MCkmUYWP76X4mTZuq7pH6MUP2tJs8",
    "attributes": Array [
      Object {
        "trait_type": "Pot",
        "value": "Yellow Plastic",
      },
      Object {
        "trait_type": "Pet",
        "value": "None",
      },
      Object {
        "trait_type": "Bark",
        "value": "Birch",
      },
      Object {
        "trait_type": "Leaves",
        "value": "Momo Maple",
      },
      Object {
        "trait_type": "Fruit",
        "value": "None",
      },
      Object {
        "trait_type": "Number of fruit",
        "value": "0",
      },
      Object {
        "trait_type": "Ground cover",
        "value": "None",
      },
      Object {
        "trait_type": "Background",
        "value": "day",
      },
    ],
    "description": "[Image (.png)](https://ipfs.io/ipfs/QmSckSnuJUwn74ZFvmNMob9dTwfZzjnXVyeKxuDYah93hx)

[Animation (.mp4)](https://ipfs.io/ipfs/QmWS953fEQgQo77n7MCkmUYWP76X4mTZuq7pH6MUP2tJs8)

[3D Model (.glb)](https://ipfs.io/ipfs/QmS8VGJL1P9kcGb2JNkPyUesYamCLRDPBg8aVuCr9qCDZM)

[AR Object (.usdz)](https://ipfs.io/ipfs/Qmd6rodUdPr3EZ2SfZ2DxjukQUjz5WaSHLBkrjAnW6yH83)
",
    "image": "https://ipfs.io/ipfs/QmSckSnuJUwn74ZFvmNMob9dTwfZzjnXVyeKxuDYah93hx",
    "name": "Bonsai #100",
  },
  "name": "Bonsai #100",
  "networkId": 1,
  "tokenAddress": "0xeC9C519D49856Fd2f8133A0741B4dbE002cE211b",
  "tokenId": "100",
  "tokenURI": "https://ipfs.io/ipfs/QmVtJUJrjWfthrGFvkwFciZcvoNLzCrC6EXAQwimYBUZeQ/100",
  "tokenURL": "https://ipfs.io/ipfs/QmVtJUJrjWfthrGFvkwFciZcvoNLzCrC6EXAQwimYBUZeQ/100",
  "tokenURLMimeType": "application/json",
}
`;
